Add Assignments Filler

Silderoy opened the pull request:
For the mighty JJ to review and ban me from the world of github forever.
- Runs in the background by JJ powers (venv)
- checks the name of the contract as written in the spreadsheet then keeps checking every cycle if it changes. if it does - runs Filler.
- checks the comments of the contract on the first cycle, the keeps checking them every cycle for new or updated comments. if it finds any - runs Filler (updated comments does not work ATM for unknown reasons, only new comments trigger the filler).
- the Filler - reads the DB and fills all the "Yes" comments into the table.
- It also cleans comment remains from the last contract.
- keeps the comments consistent through updates and changes to attendance.
- marks recruits as what they are.

TyroneMF commented on server/Assignments_filler/‏‏DB_call.py:
```diff
@@ -0,0 +1,73 @@
+        logging.debug("Query executed successfully: ** %s " % (query.replace("\n", " ")))
+    except mysql.connector.Error as error :
+        connection.rollback() #rollback if any exception occured
+        logging.error("Failed performing query {}".format(error))
```
```suggestion
        logging.error(f"Failed performing query {error}")
```

Switch to f strings instead of using .format

TyroneMF commented on server/Assignments_filler/service_nssm.log:
```diff
@@ -0,0 +1,1236 @@
+
```
Don't need log files!

TyroneMF commented on server/Assignments_filler/run_windows.bat:
```diff
@@ -0,0 +1 @@
+venv\Scripts\activate.bat && python sheetFiller.py && deactivate
```
Empty line at end of file

jonpas commented on server/Assignments_filler/requirements.txt:
```diff
@@ -0,0 +1,27 @@
+uritemplate==3.0.1
+urllib3==1.26.4
+
+Also requires "credentials.json" from the google API project page.
```
This file is generated with `pip freeze > requirements.txt` and defines specific requirements for a virtual environment so it can always be ran with same exact libraries.

Next time requirements change or get updated, this line will be gone. Put it into a `README.md` instead along with all other necessary instructions.

jonpas replied:
Not resolved yet.

jonpas replied:
Stop marking this as solved, it was not. Regenerate the file as I wrote above.

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+from __future__ import print_function
```
Remove, we are on Python 3, you don't need a print function, you already have it.

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+from __future__ import print_function
+
+import os.path
```
```suggestion
import os
```
You use it as `os.path`, just import the module in its entirety.

jonpas commented on server/Assignments_filler/‏‏DB_call.py:
```diff
@@ -0,0 +1,73 @@
+#region imports
```
```suggestion
```

jonpas commented on server/Assignments_filler/‏‏DB_call.py:
```diff
@@ -0,0 +1,73 @@
+from mysql.connector import Error
+from mysql.connector import errorcode
+
+import logging
+import os
```
Put above MySQL, follow this format:
```
import [default Python modules]

import [PyPI/libraries/external]

import [other local modules]
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s]: %(message)s",
+                    handlers=[logging.FileHandler("filler.log"), logging.StreamHandler()])
+
+def readSheet(service, range, amount = 1):
```
```suggestion
def readSheet(service, range, amount=1):
```
Arguments have no spaces between name and value.

jonpas replied:
Why did you revert this!?

Silderoy replied:
By mistake. It's already fixed

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+    readRange = TEAM_ASSINGMENT_TAB+RANGE_READ
+    oldContractName = contractName
+    oldNID = nid
+    query = ("SELECT `created` FROM drupal.comment where nid=%s ORDER BY created DESC" % (oldNID))
```
This is deprecated Python 2 formatting, use f-strings.

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+    checker(service, contractName, nid, numberOfComments)
+
+if __name__ == '__main__':
+    main()
```
Missing empty line at end of file (EOF). This is suggested for Git diffs to work properly and reduce merge conflicts.

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+    
+    logging.info("Successfully filled!")
+    
+    checker(service, contractName, nid, numberOfComments)
```
Checker calls Filler, Filler calls checker = cyclic dependency.

Avoid that, Checker can call Filler and that's it. It simply continues checking after that.

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+
+def main():
+    #Setting up the API service
+    #region
```
What is this region stuff?

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+        goodListOfNames.extend(sublist)
+    commentsRange = TEAM_ASSINGMENT_TAB + "D2:D" + str(numberOfComments+1)
+    listOfComments = readSheet(service, commentsRange, 2)
+    print (goodListOfNames)
```
```suggestion
    print(goodListOfNames)
```
No space between function name and parentheses.

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,237 @@
+    readRange = TEAM_ASSINGMENT_TAB+RANGE_READ
+    
+    #read new contract name from DB
+    writtenName = readSheet(service, readRange)
+    query = ("SELECT `nid`,`title` FROM drupal.node where title like \"%%%s%%\"" % (writtenName))
+    contractList = get_from_db(query)
+    nid = contractList[0][0]
+    contractName = contractList[0][1]
+    query = ("SELECT `created` FROM drupal.comment where nid=%s ORDER BY created DESC" % (nid))
+    commentList = get_from_db(query)
+    numberOfComments = len(commentList)
```
Exact same code as in Checker function? Why?

jonpas commented on server/Assignments_filler/‏‏DB_call.py:
```diff
@@ -0,0 +1,73 @@
+import logging
+import os
+
+# from dotenv import load_dotenv
+# load_dotenv()
+#endregion
```
What for?

jonpas commented on server/Assignments_filler/‏‏DB_call.py:
```diff
@@ -0,0 +1,73 @@
+        return [["Currapted query"]]
+
+    try:
+        connection = mysql.connector.connect(host=db_host, database=db_name, port=db_port, user=db_user, password=db_pass)
```
This will establish a fresh MySQL connection each time it runs, and it can run multiple times because you do multiple `get_from_db` calls in a short amount of time.

Connect once on start and just use the connection.

jonpas commented on server/Assignments_filler/‏‏DB_call.py:
```diff
@@ -0,0 +1,73 @@
+        logging.error("Failed performing query {}".format(error))
+        return str(error)
+    finally:
+        if(connection.is_connected()):
```
```suggestion
        if connection.is_connected():
```
Redudant parentheses.

jonpas reviewed (changes requested):
Any comment in this review does **not** apply **only for** the given line it is on!

I also suggest you download `flake8` linter (for whatever editor you have) and use it (change max line length to 120).

jonpas commented:
Also rename the folder to `assignments_filler`.

Silderoy commented:
files updated for DB connection astablishment note

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,216 @@
+import time
+
+import DB_call
+from DB_call import get_from_db
```
`DB_call` imports in the third "import group". You also don't need standalone `DB_call`.

jonpas replied:
Scratch that, you do, because you start the connection that way. A possibly better way to do this would be to have an `init` call in `DB_call.py`, that you run once in `sheetFiller.py` before starting the `checker`. Then you can just import the 2 functions you use and keep it nicely separated.

Silderoy replied:
I don't understand that one.
Do you mean I don't need line 7 since line 6 already covers it?
if you do - I know, just makes it a bit shorter overall.

jonpas replied:
No, I mean you shouldn't need line 6. :)

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,216 @@
+    readRange = TEAM_ASSINGMENT_TAB+RANGE_READ
+    
+    while True:
+        checker(service)
```
```suggestion
    try:
        checker(service)
    except (KeyboardInterrupt, SystemExit):
        return
```
To handle shutting down gracefully instead of throwing an error.

jonpas commented on server/Assignments_filler/‏‏DB_call.py:
```diff
@@ -0,0 +1,50 @@
+                    handlers=[logging.FileHandler("filler.log"), logging.StreamHandler()])
+
+def connect_db():
+    return mysql.connector.connect(host=db_host, database=db_name, port=db_port, user=db_user, password=db_pass)
```
This should resolve the cache issue. In MySQL if you don't commit, next time you run the same query, it will give you the same result.

Doesn't seem to fix the issue at hand though, I think there is something with comparing if there is a new comment broken as well.
```suggestion
    connection = mysql.connector.connect(host=db_host, database=db_name, port=db_port, user=db_user, password=db_pass)
    connection.autocommit = True
    return connection
```

TyroneMF commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,219 @@
+import os
```
```suggestion
import os, logging, sys, time
```

jonpas replied:
No

TyroneMF commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,219 @@
+import os
+import logging
```
```suggestion

```

jonpas replied:
No

TyroneMF commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,219 @@
+import os
+import logging
+import sys
```
```suggestion

```

jonpas replied:
No

TyroneMF commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,219 @@
+import os
+import logging
+import sys
+import time
```
```suggestion

```

jonpas replied:
No

Silderoy commented:
Added a few features:
1. if you write a typo/non-existant contract name - the cell will be marked in red (and the script will keep running without an error)
once you fix it the cell changes back to black borders.
could not make it modular to any cell so its fixed on the right cell.

2. added system to check if more than one contract fits the written name (for example the "V2" contracts we have this saturday).
Once it happens - it checks if any of them fits the exact name (so if you write the exact contract name it will choose it, even in another contract contains the name within its own (see "sand spider" and "sand spider V2").
If there is no exact match - it will take the newest.

3. Log file now gets a custom log that fits the situation (found X contracts, chose the exact/the latest).

4. Fixed error if a contract has no responses when its entered.

Also made sure all of that has minimal impact on the DB and the Sheet, so it doesn't spam them.

If you want to give it a check

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+        valueInputOption='USER_ENTERED', body=body).execute()
+        
+def paintSheet(service, color):
+    if (color == "red"):
```
```suggestion
    if color == "red":
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+def paintSheet(service, color):
+    if (color == "red"):
+        redLevel = 0.9
+    elif (color == "black"):
```
```suggestion
    elif color == "black":
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+    if (queryLength == 0):
+        oldNID = 0
+        oldContractName = "empty name"
+    elif (queryLength > 1):
```
```suggestion
    elif queryLength > 1:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+    elif (queryLength > 1):
+        for sublist in contractList:
+            exactName = sublist[1]
+            if (writtenName == exactName):
```
```suggestion
            if writtenName == exactName:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+    query = (f"SELECT `created` FROM drupal.comment where nid={oldNID} ORDER BY created DESC")
+    oldCommentList = get_from_db(query)
+    oldNumberOfComments = len(oldCommentList)
+    if (oldNumberOfComments):
+        oldLastUpdated = oldCommentList[0][0]
+    else:
+        oldLastUpdated = 1
```
```suggestion
    oldLastUpdated = 1
    if oldNumberOfComments:
        oldLastUpdated = oldCommentList[0][0]
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+        query = (f"SELECT `nid`,`title` FROM drupal.node where title like \"%{writtenName}%\" ORDER BY nid DESC")
+        contractList = get_from_db(query)
+        queryLength = len(contractList)
+        if (queryLength == 0):
```
```suggestion
        if queryLength == 0:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+        contractList = get_from_db(query)
+        queryLength = len(contractList)
+        if (queryLength == 0):
+            if (not isRed):
```
```suggestion
            if not isRed:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+                isRed = True
+            time.sleep(refreshRate)
+            continue
+        elif (queryLength > 1):
```
```suggestion
        elif queryLength > 1:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+        elif (queryLength > 1):
+            for sublist in contractList:
+                exactName = sublist[1]
+                if (writtenName == exactName):
```
```suggestion
                if writtenName == exactName:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+        query = (f"SELECT `created` FROM drupal.comment where nid={newNID} ORDER BY created DESC")
+        commentList = get_from_db(query)
+        newNumberOfComments = len(commentList)
+        if (newNumberOfComments):
+            newLastUpdated = commentList[0][0]
+        else:
+            newLastUpdated = 1
```
```suggestion
        newLastUpdated = 1
        if newNumberOfComments:
            newLastUpdated = commentList[0][0]
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+            newLastUpdated = 1
+        
+        #compare the new and the old data
+        if (newContractName != oldContractName):
```
```suggestion
        if newContractName != oldContractName:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+            logging.info(f"New contract name detected! {nameLog}")
+            nameChanged = True
+            break
+        elif (newNumberOfComments != oldNumberOfComments):
```
```suggestion
        elif newNumberOfComments != oldNumberOfComments:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+            logging.info("A comment was added!")
+            nameChanged = False
+            break
+        elif (newLastUpdated != oldLastUpdated):
```
```suggestion
        elif newLastUpdated != oldLastUpdated:
```

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+            nameChanged = False
+            break
+        time.sleep(refreshRate)
+    filler(service, nameChanged, isRed, newContractName, newNID)
```
You should define a default value for `nameChanged`, it might not be set as well. Won't error because `bool(None)` is still `False`, but good practice.

jonpas commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+    #fill mission full name
+    if nameChanged:
+        writeSheet(service, contractName, readRange, 1)
+        if (isRed):
```
```suggestion
        if isRed:
```
This is Python!

Silderoy commented on server/Assignments_filler/sheetFiller.py:
```diff
@@ -0,0 +1,302 @@
+    body = {"requests": [{
+        "updateBorders": {
+            "range": { #which cell will get formatted
+                "sheetId": 871598544, #the right tab
```
I couldn't figure out a way to turn the "F33" format into this range format..

Silderoy commented:
merged most changes + added explenations inside the file for different sections + small fixes

jonpas commented:
Replaced by https://github.com/Theseus-Aegis/Tools/blob/3402b4a672c606dcf4d94f9ab3631557eb733dec/gappsscript/assignments/fillAssignments.gs

API from Athena will eventually be part of Athena2, so this also simplifies that transition.
